src/parser/queries/QueryBuilder.ts

```diff
@@ -36,25 +42,57 @@ export class QueryBuilder {
 
         if (this.#language !== Language.Unknown) {
             if (this.#statements.length > 0) {
-                queryString = this.getStatementsQuery();
+                queryString = this.#getStatementsQuery();
             }
 
             dlog("------------- Start Query: --------------");
             dlog(queryString);
             dlog("-----------------------------------------");
 
-            return new Query(this.#treeSitterLanguage, queryString);
+            return this.#retrieveTreeSitterQuery(queryString);
         }
     }
 
+    /**
+     * Tries to retrieve the tree-sitter query from the cache. If there is no matching query,
+     * calls the constructor of tree-sitter.Parser.Query, which is a very performance-heavy operation.
+     */
+    #retrieveTreeSitterQuery(queryString: string): Query {
+        let cachedQueriesForLanguage = treeSitterQueryCache.get(this.#language);
+
+        if (cachedQueriesForLanguage !== undefined) {
+            const cachedQuery = cachedQueriesForLanguage.get(queryString);
+            if (
+                cachedQuery !== undefined &&
+                cachedQuery.queryString === queryString &&
+                cachedQuery.language === this.#language
+            ) {
+                return cachedQuery.query;
+            }
+        }
+        const newQuery = new Query(languageToGrammar.get(this.#language), queryString);
+
+        if (cachedQueriesForLanguage === undefined) {
+            cachedQueriesForLanguage = new Map<string, CachedQuery>();
+            treeSitterQueryCache.set(this.#language, cachedQueriesForLanguage);
+        }
+        treeSitterQueryCache.get(this.#language)?.set(queryString, {
+            language: this.#language,
+            queryString: queryString,
+            query: newQuery,
+        });
```

Comment:
I think when we remove CachedQuery, the code could be more readable. What do you think?
```suggestion
        let queriesForLanguage = treeSitterQueryCache.get(this.#language);
        const query = queriesForLanguage?.get(queryString);

        if (queriesForLanguage !== undefined && query !== undefined)  return query;

        const newQuery = new Query(languageToGrammar.get(this.#language), queryString);

        if (queriesForLanguage === undefined) {
            treeSitterQueryCache.set(this.#language, new Map<string, Query>());
        }
        treeSitterQueryCache.get(this.#language)?.set(queryString, newQuery);
        return newQuery;
```

Reply:
I have done something similar now